### DISCUSSION TITLE
Any way to detect HTTP errors with pyodide.http.open_url()?

### DISCUSSION THREAD

**Discussion by whitphx** [pyodide/pyodide#3115] [Q&A] [answered]
`pyodide.http.open_url()` does not throw an error even when the status code is not 2xx and it simply returns the response body like the normal.
Are there any way to detect the error status with `pyodide.http.open_url()`?

I'm have tested it with Pyodide v0.21.2.

**Comment by ryanking13** [comment, marked as answer]
The implementation of `open_url` is very simple, so you can extend its functionalities by re-implementing it.

```py
def open_url(url: str) -> StringIO:
    req = XMLHttpRequest.new()
    req.open("GET", url, False)
    req.send(None)
    return StringIO(req.response)
```

Though our recommendation is to use [`pyfetch`](https://pyodide.org/en/stable/usage/api/python-api/http.html#pyodide.http.pyfetch) instead of `open_url`.

**Reply by whitphx** [reply]
Thank you!

**Reply by rth** [reply]
I think it wouldn't hurt to fix it on our end as well, a lot of applications need to have a sync way of fetching files and `pyfetch` is async.

BTW, related interesting projects that monkeypatch requests (with essentially something like open_url) https://github.com/koenvo/pyodide-http and https://github.com/emscripten-forge/requests-wasm-polyfill

**Reply by whitphx** [reply]
Thank you.

> a lot of applications need to have a sync way of fetching files and pyfetch is async.

Yes, my case was exactly one of them.
As @ryanking13 said, the implementation is so simple that extending it by ourselves is easy, but I think it would also be better if the official implementation is provided to avoid confusion. I respect the project decision though (It also makes sense if the project officially does not pay efforts on `open_url` since `pyfetch` is recommended).

The reference projects are also interesting. Thank you :)
